### ISSUE TITLE
[Feature request] Highlight character differences between an unbound module/value/... and its hint

### ISSUE THREAD

**Issue by kit-ty-kate** [ocaml/ocaml#11927]
Take for example the following error message:
```
Unbound module FullTypeDefinition
Hint: Did you mean FullTypeDefiniton?
```
A keen eye might see the typo, but the difference blends fairly well that some people might get confused and suppose there is a bug in the compiler or something of the sort and get frustrated.

To make it more obvious what the issue actually is, i think it might be interesting to highlight where the differences are.

For example, in this case, the first line could have the final `i` character in bold letter or red to highlight where the issue is.

**Comment by gasche** [comment]
It would indeed be nice to underline the characters that are different on both sides. Another approach, which would be necessary on "dumb" terminals, would be to print the diff explicitly:

```
Unbound module FullTypeDefinithon
Hint: Did you mean FullTypeDefinition?
  FullTypeDefinit[-h][+i]on
```

To do either of these things we could use the utils/Diffing module that can compute a "diff" between two lists of elements of arbitrary type (here lists of `Uchar.t` I guess).

Now that I think of it: doing this well for unicode characters is going to be very tricky. Maybe a good-enough approach is to group added/removed chars together (foo[+bar] instead of foo[+b][+a][+r]), which maximizes the chances that we actually display the higher-level units in a sensible way.

**Comment by dbuenzli** [comment]
A first simpler step would be to align the tokens to make it easier for the eye to diff (no unicode problem here since what comes before should always be pure ascii):
```
      Unbound module FullTypeDefinithon
Hint:   Did you mean FullTypeDefinition?
```

**Comment by gasche** [comment]
Good idea. We might be able to do something with Format tabulation boxes (but maybe those are deprecated now), or maybe we should put both lines on their own lines.

**Comment by shindere** [comment]
Daniel Bünzli (2023/01/22 08:04 -0800):
> A first simpler step would be to align the tokens to make it easier for the eye to diff (no unicode problem here since what comes before should always be pure ascii):
> ```
>       Unbound module FullTypeDefinithon
> Hint:   Did you mean FullTypeDefinition?
> ```

For the eye but not for the fingers :-)

I really liked theidea ofhaving thediff shown in a linearway I have to
say because I found the initial example quite convincing -- I really had
a hard timie figuring out what the diff was and I even found something
different fromwhat @gashc wrote: he wrote that an h had been typed
instead of an i, but I read that an i had been forgotten.

**Comment by gasche** [comment]
In any cases, the example in #11929 (which I wrote while trying to implement @dbuenzli's suggestion) show that it is not possible to perform this kind of vertical alignment in many cases, as the position of the identifiers in the error message can be very different depending on the message. I think that the textual diff will work better.

**Comment by shindere** [comment]
Gabriel Scherer (2023/02/08 14:26 -0800):
> In any cases, the example in #11929 (which I wrote while trying to
> implement @dbuenzli's suggestion) show that it is not possible to
> perform this kind of vertical alignment in many cases, as the position
> of the identifiers in the error message can be very different
> depending on the message. I think that the textual diff will work
> better.

That's a news I very much like!!:)

**Comment by OlivierNicole** [comment]
Commenting to keep this open

**Comment by gasche** [comment]
Since #13837, the output of the "Did you mean" hints is now aligned:

```
# FullTypeDefiniton.x;;
Error: Unbound module FullTypeDefiniton
Hint:    Did you mean FullTypeDefinition?
```

I propose to close the issue as I think this is a nice (partial) solution to the problem.
